ospfd/ospf_gr.c

```diff
@@ -0,0 +1,824 @@
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; see the file COPYING; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+ */
+
+#include <zebra.h>
+
+#include "memory.h"
+#include "command.h"
+#include "table.h"
+#include "vty.h"
+#include "log.h"
+#include "printfrr.h"
+
+#include "ospfd/ospfd.h"
+#include "ospfd/ospf_abr.h"
+#include "ospfd/ospf_flood.h"
+#include "ospfd/ospf_ism.h"
+#include "ospfd/ospf_interface.h"
+#include "ospfd/ospf_asbr.h"
+#include "ospfd/ospf_lsa.h"
+#include "ospfd/ospf_route.h"
+#include "ospfd/ospf_zebra.h"
+#include "ospfd/ospf_lsdb.h"
+#include "ospfd/ospf_neighbor.h"
+#include "ospfd/ospf_opaque.h"
+#include "ospfd/ospf_nsm.h"
+#include "ospfd/ospf_gr.h"
+#include "ospfd/ospf_errors.h"
+#include "ospfd/ospf_dump.h"
+
+static void ospf_gr_nvm_delete(struct ospf *ospf);
+
+/* Lookup self-originated Grace-LSA in the LSDB. */
+static struct ospf_lsa *ospf_gr_lsa_lookup(struct ospf *ospf,
+					   struct ospf_area *area)
+{
+	struct ospf_lsa *lsa;
+	struct in_addr lsa_id;
+	uint32_t lsa_id_host_byte_order;
+
+	lsa_id_host_byte_order = SET_OPAQUE_LSID(OPAQUE_TYPE_GRACE_LSA, 0);
+	lsa_id.s_addr = htonl(lsa_id_host_byte_order);
+	lsa = ospf_lsa_lookup(ospf, area, OSPF_OPAQUE_LINK_LSA, lsa_id,
+			      ospf->router_id);
+
+	return lsa;
+}
+
+/* Fill in fields of the Grace-LSA that is being originated. */
+static void ospf_gr_lsa_body_set(struct ospf_gr_info *gr_info,
+				 struct ospf_interface *oi, struct stream *s)
+{
+	struct grace_tlv_graceperiod tlv_period = {};
+	struct grace_tlv_restart_reason tlv_reason = {};
+	struct grace_tlv_restart_addr tlv_address = {};
+
+	/* Put grace period. */
+	tlv_period.header.type = htons(GRACE_PERIOD_TYPE);
+	tlv_period.header.length = htons(GRACE_PERIOD_LENGTH);
+	tlv_period.interval = htonl(gr_info->grace_period);
+	stream_put(s, &tlv_period, sizeof(tlv_period));
+
+	/* Put restart reason. */
+	tlv_reason.header.type = htons(RESTART_REASON_TYPE);
+	tlv_reason.header.length = htons(RESTART_REASON_LENGTH);
+	if (gr_info->restart_support)
+		tlv_reason.reason = OSPF_GR_SW_RESTART;
+	else
+		tlv_reason.reason = OSPF_GR_UNKNOWN_RESTART;
+	stream_put(s, &tlv_reason, sizeof(tlv_reason));
+
+	/* Put IP address. */
+	if (oi->type == OSPF_IFTYPE_BROADCAST || oi->type == OSPF_IFTYPE_NBMA
+	    || oi->type == OSPF_IFTYPE_POINTOMULTIPOINT) {
+		tlv_address.header.type = htons(RESTARTER_IP_ADDR_TYPE);
+		tlv_address.header.length = htons(RESTARTER_IP_ADDR_LEN);
+		tlv_address.addr = oi->address->u.prefix4;
+		stream_put(s, &tlv_address, sizeof(tlv_address));
+	}
```

Comment:
IP address must be advertised Routerid for other  link types . like point to point.

Reply:
> What about pain to point case , IP address must be advertised Routerid for other link types , right ?

The RFC says the following:
> A.  Grace-LSA Format
> [snip]
>   o  IP interface address (Type=3, length=4).  The router's IP
>      interface address on the subnet associated with the grace-LSA.
>      Required on broadcast, NBMA and Point-to-MultiPoint segments,
>      where the helper uses the IP interface address to identify the
>      restarting router (see Section 3.1).

And:
> 3.1.  Entering Helper Mode
> [snip]
>      1) Y currently has a full adjacency with X (neighbor state Full)
>         over the associated network segment.  On broadcast, NBMA and
>         Point-to-MultiPoint segments, the neighbor relationship with X
>         is identified by the IP interface address in the body of the
>         grace-LSA (see Appendix A).  On all other segment types, X is
>         identified by the grace-LSA's Advertising Router field.

So it seems this TLV isn't necessary for point-to-point interfaces.

The GR helper code handles that correctly:
```c
        if (oi->type != OSPF_IFTYPE_POINTOPOINT) {
                restarter = ospf_nbr_lookup_by_addr(oi->nbrs, &restart_addr);

                if (!restarter) {
                        if (IS_DEBUG_OSPF_GR)
                                zlog_debug(
                                        "%s, Restarter is not a nbr(%pI4) for this router.",
                                        __func__, &restart_addr);
                        return OSPF_GR_NOT_HELPER;
                }
        } else
                restarter = nbr;
```